src/EditorFeatures/Core/Implementation/ForegroundNotification/ForegroundNotificationService.cs

```diff
@@ -139,6 +147,15 @@ private void NotifyOnForegroundWorker()
                         {
                             // eat up cancellation
                         }
+                        catch (Exception ex) when (FatalError.ReportWithoutCrash(ex))
+                        {
+                            // The PendingWork callbacks should never throw.  In the case they do we
+                            // must ensure the IAsyncToken implementation is completed.  If it is not
```

Comment:
@jasonmalinowski this is the IDE code. My change is a representation of how I believe it should function given the comments and current structure.  If you're asking on how I think it should work then look to this change for guidance.  If you'd prefer it be done another way then please be more definitive.